help: Provide clearer description of NotificationCenter mechanism

iani opened the pull request:
This is not MVC but the Observer pattern.
The description of the different components model, message, dependant,
action and of the two steps register/notify is made more explicit here.

<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation

Describe the components and function of the NotificationCenter mechanism
more explicitly, to clarify how it is used.

<!-- If this fixes an open issue, link to it by writing "Fixes #555." -->

## Types of changes

Substituted Observer Pattern instead of Model-View-Controller pattern in the initial description, because Observer Pattern corresponds more accurately to what is described. 

Added description of the update mechanism and its parts, as well as discussion of the motivation for using this mechanism over the simple Observer mechanism.

- Documentation
## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

telephon commented on HelpSource/Classes/NotificationCenter.schelp:
```diff
@@ -4,9 +4,35 @@ related:: Classes/SimpleController, Classes/NodeWatcher
 
 description::
-One common OOP pattern is Model-View-Controller where one object (the controller) is a dependant of the model. Every time the model changes it notifies all of its dependants. In this case the model has a dictionary of dependants and iterates through those.
+One common OOP pattern is the Observer Pattern where one object (the controller or observer) is a dependant of a model or subject object ( https://en.wikipedia.org/wiki/Observer_pattern ). Every time the model changes it notifies all of its dependants. In the default implementation of this pattern in SuperCollider, the model has a dictionary of dependants and iterates through those.  A limitation ...
```
suggestion: –>

If a dependant is required to perform different actions depending on the type of change that happens in the model, or on the dependant itself, then this must be specified in the update method of the dependant.

iani replied:
Hello Julian, I have integrated your suggestion in the help file of NotificationCenter and pushed the commit to my fork of supercollider git repo.  The commit is 933c52ee582e823c6dda3204cfbb1c8d24cf12b8 

I do not know how to send this commit to the supercollider/supercollider repo without opening another pull request. I will ask advice from Nathan who is maintaining the instructions for contributions on the wiki.

snappizz commented:
hi @iani, here's how to fix your branch. first make a backup just in case:

    git checkout NotificationCenter/better-help
    git branch NotificationCenter/better-help/backup

then roll back your current branch (not the backup) to the last good commit:

    git reset --hard ca5ba03

and cherry-pick the "integrate suggestion by JR" commit, skipping over the unintended merge  commit:

    git cherry-pick 933c52e

and force push:

    git push -f origin NotificationCenter/better-help

let me know if you run into any issues.

iani commented:
> let me know if you run into any issues.

Hi @snappizz I did the reset and cherrypick and the suggestion by @telephon is now integrated in the pull request. Waiting for the next steps.

snappizz commented on HelpSource/Classes/NotificationCenter.schelp:
```diff
@@ -4,9 +4,35 @@ related:: Classes/SimpleController, Classes/NodeWatcher
 
 description::
-One common OOP pattern is Model-View-Controller where one object (the controller) is a dependant of the model. Every time the model changes it notifies all of its dependants. In this case the model has a dictionary of dependants and iterates through those.
+One common OOP pattern is the Observer Pattern where one object (the controller or observer) is a dependant of a model or subject object ( https://en.wikipedia.org/wiki/Observer_pattern ). Every time the model changes it notifies all of its dependants. In the default implementation of this pattern in SuperCollider, the model has a dictionary of dependants and iterates through those.  A limitation ...
```
i think this could be easier to read as multiple paragraphs.

snappizz replied:
also, there seems to be considerable confusion in this help file due to use of multiple names for the same thing. it's model/dependant in the help file, but object/listener in the actual argument names. i'd suggest always writing "object" and "listener" to keep it consistent.

jrsurge replied:
GOF book uses `Observer` and `Subject`, would be good to use their terms?

snappizz replied:
imo it would be confusing to use one convention in the help file and a different one in the actual API, which is "object" and "listener". it's unfortunate that the API seems to deviate from standard terminology (and can't be changed due to backward compatibility) but it's most important to be internally consistent.

iani replied:
Thanks for the comments. I will split explanation to multiple paragraphs and rename the argument names for consistency as suggested. I agree with @snappizz that internal consistency has priority. References to GOF terminology can be provided in the present help documentation, and I will try to do this.

snappizz commented on HelpSource/Classes/NotificationCenter.schelp:
```diff
@@ -4,9 +4,35 @@ related:: Classes/SimpleController, Classes/NodeWatcher
+## The model (can be any type of object).
+## The notification message (usually a Symbol).
+## The dependant (can be any type of object).
+## The action to be performed (usually a Function).
```
what arguments is the function called with?

iani replied:
The four items provided here are the arguments. I will make that explicit in an example.

snappizz replied:
let me say it a different way -- does the action receive any arguments when it is triggered?

snappizz commented on HelpSource/Classes/NotificationCenter.schelp:
```diff
@@ -4,9 +4,35 @@ related:: Classes/SimpleController, Classes/NodeWatcher
 
-Another common pattern is NotificationCenter wherein an object emits a notification and clients can register functions that will be executed when that notification happens.
+NUMBEREDLIST::
+## The model (can be any type of object).
```
`nil` seems to be an exception to this. `NotificationCenter.register(nil, \foo, \a, { });` gives me an error. possible bug?

iani replied:
OK, so how about: 
``` ## the model (can be any type of object except nil) ```
?

iani commented:
The questions posed by @snappizz and @jrsurge have now been answered and are available in commit https://github.com/supercollider/supercollider/commit/d178054e8c86b71dcef8c57c345247ae62efdfd5

iani commented:
PS: I added some usage scenario examples at the beginning of the help file, to explain the usefulness of this pattern more clearly. See commit https://github.com/supercollider/supercollider/pull/4494/commits/24c9b842b22b8603ceb1f66cef3c62189931c737

brianlheim commented on HelpSource/Classes/NotificationCenter.schelp:
```diff
@@ -4,26 +4,60 @@ related:: Classes/SimpleController, Classes/NodeWatcher
+code::
+// using default server as subject and a symbol as observer:
+NotificationCenter.register(Server.default, \newAllocators, \yourself, {
+    // Substitute anything more meaningful here:
```
please use tabs for indentation in code examples

brianlheim commented on HelpSource/Classes/NotificationCenter.schelp:
```diff
@@ -4,26 +4,60 @@ related:: Classes/SimpleController, Classes/NodeWatcher
+NUMBEREDLIST::
+## A subject adds a dependant with: "subject.addDependant(observer)"
+## When a subject wants to inform its observers (dependants) that it has changed, it uses the method changed: "subject.changed(<optional arguments>...)"
+## The method changed obtains the list of dependants of the subject and iterates over each observer sending it the message "update": "observer.changed(subject, <optional arguments>...)"
```
you can use `code:: receiver.message(a, b, c) ::` for inline code, please change these to use that

iani commented:
Hello Brian,

Thank you for the comments.
I will look at this and submit a commit taking your suggestions into
consideration.
Best,
Iannis

On Mon, Dec 23, 2019 at 12:42 AM Brian Heim <notifications@github.com>
wrote:

> *@brianlheim* requested changes on this pull request.
>
> just a few minor things, sorry for the delay in reviewing. @snappizz
> <https://github.com/snappizz> any thoughts on this? you still have a
> pending changes requested review
> ------------------------------
>
> In HelpSource/Classes/NotificationCenter.schelp
> <https://github.com/supercollider/supercollider/pull/4494#discussion_r360733333>
> :
>
> >  ::
>
> -The link::Classes/Buffer:: class register a function to clear its info cache whenever a server restarts. The server is emitting changed messages quite often (every 0.4 secs for the status updates), and the Buffer class is only interested in boot/quit events, so this is a more lightweight system for this purpose.
> +Therefore any object that wants to perform an action when a server boots or quits, can use NotificationCenter to register the action using the following code:
> +
> +code::
> +// using default server as subject and a symbol as observer:
> +NotificationCenter.register(Server.default, \newAllocators, \yourself, {
> +    // Substitute anything more meaningful here:
>
> please use tabs for indentation in code examples
> ------------------------------
>
> In HelpSource/Classes/NotificationCenter.schelp
> <https://github.com/supercollider/supercollider/pull/4494#discussion_r360733386>
> :
>
> >
> -A link::Classes/Server:: emits a \newAllocators notification when it creates new node and bus allocators which it does when it quits or boots.
> +The implementation provided by NotificationPattern is a variant of a common OOP pattern called the Observer Pattern, which provides a way for an object to flexibly broadcast messages to interested receiver objects.  This pattern defines a one-to-many dependency between objects so that when one object changes state, all its dependents are notified and updated automatically.  (see: https://www.gofpatterns.com/behavioral-design-patterns/behavioral-patterns/observer-pattern.php and https://en.wikipedia.org/wiki/Observer_pattern ).  An object that  notifies its dependents is called the subject, and the dependents are called observers.
> +
> +The original implementation of this pattern in SuperCollider involves three steps:
> +
> +NUMBEREDLIST::
> +## A subject adds a dependant with: "subject.addDependant(observer)"
> +## When a subject wants to inform its observers (dependants) that it has changed, it uses the method changed: "subject.changed(<optional arguments>...)"
> +## The method changed obtains the list of dependants of the subject and iterates over each observer sending it the message "update": "observer.changed(subject, <optional arguments>...)"
>
> you can use code:: receiver.message(a, b, c) :: for inline code, please
> change these to use that
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/supercollider/supercollider/pull/4494?email_source=notifications&email_token=AABMNRAXRWLXONJ3RTVQKS3QZ7UNFA5CNFSM4IFH35BKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQA7WXA#pullrequestreview-335674204>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABMNRGXOS3V2CUDDGWW5CTQZ7UNFANCNFSM4IFH35BA>
> .
>

brianlheim commented:
thanks Iannis!

brianlheim commented:
closing as stale, please reopen if you want to update this @iani
